gnrc_netif: Add support for internal event loop

gebart opened the pull request:
### Contribution description

Enabled by the `gnrc_netif_events` pseudo module. Using an internal event loop eliminates the risk of lost interrupts and lets ISR events always be handled before any send/receive requests from other threads are processed.
The events in the event loop is also a potential hook for MAC layers and other link layer modules which may need a way to inject and process events before any external IPC messages are handled.

### Issues/PRs references

miri64 commented on sys/include/net/gnrc/netif.h:
```diff
@@ -56,6 +59,16 @@ extern "C" {
+typedef struct {
+    event_t super;
+    netdev_t *dev;
+} event_netdev_t;
```
Module prefix missing. (I'd rather prefer the name `gnrc_netif_event_t` then ;-))

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -36,6 +36,12 @@
+#ifdef MODULE_GNRC_NETIF_ASYNC_EVENTS
+typedef struct {
+    event_netdev_t isr;
+} event_table_t;
```
I prefer a leading `_` here, to easily identify this as a c-file internal type.

miri64 commented on sys/include/net/gnrc/netif.h:
```diff
@@ -75,6 +88,21 @@ typedef struct {
+     * This pointer is a way to pass an event instance allocated on the stack by
+     * the netif thread to the IRQ event handler called by the netdev driver
+     * while still keeping the allocation of specific events an implementation
+     * detail of gnrc_netif.c
```
I've read this sentence now three times and still find it hard to grasp it. Can you split this up into smaller sentences maybe? Also I think, if I understand it correctly, that this is more an implementation note, and should thus not be in the detailed section, but in the `@note` section.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1254,63 +1284,92 @@ static void *_gnrc_netif_thread(void *args)
     while (1) {
+        int msg_waiting = 0;
+#ifdef MODULE_GNRC_NETIF_ASYNC_EVENTS
+        /* Using messages for ordered, synchronous IPC, and events for
```
`GNRC_NETAPI_MSG_TYPE_SND` isn't synchronous, and also events aren't any less ordered or unordered than `msg_t` so I don't really get this statement...

gebart commented:
@miri64 Perhaps a bad choice of words. The idea is to make sure that ISRs will always be executed before any queued external requests (send/recv/get/set). In the future, it will also be useful to be able to insert other implementation specific events which can be handled first just like the ISR event, for example turning on the radio when it is time for a wake period in a duty cycling MAC, or handling timeouts and other exceptional events.

gebart commented:
There's also the concept that there should only ever be a need for exactly one queued ISR event in the event loop, since the radio driver should probably take care of everything that has to be done in one go and the duplicated ISR events will just be no-ops, whereas there may be several queued send events with different data.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -37,6 +37,20 @@
+
+typedef struct {
+    event_netdev_t isr;
+} gnrc_netif_events_t;
```
Why are those split up into into two types?

gebart replied:
Originally I had thought that there would be more than isr in this struct, but it showed that it was not necessary. I have removed this now and refactored the code.

gebart commented:
@miri64 rebased, refactored to address comment.
Testing:
Add USEMODULE += gnrc_netif_events to the makefile. I have added a test commit for this change in the gnrc_networking example.

gebart commented:
@miri64 The ContikiMAC implementation requires this event loop to be able to schedule waking periods and retransmissions properly, using the message queue would let applications interfere with the timing of wake periods and retransmissions, or we would have to perform tasks from within interrupt context to preempt the normal flow of the thread.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1305,63 +1337,93 @@ static void *_gnrc_netif_thread(void *args)
+        /* Only messages used for event handling */
+        DEBUG("gnrc_netif: waiting for incoming messages\n");
+        msg_waiting = msg_receive(&msg);
+#endif /* MODULE_GNRC_NETIF_EVENTS */
```
This whole `ifdef` block should IMHO go into its own function with `msg_waiting` as a return value.

miri64 replied:
(this would also make the diff more readable ;-))

gebart replied:
Addressed

miri64 commented:
I think apart from the ifdef issue pointed out above, I think I'm alright with this, but I need to see when the diff is better trackable.

gebart commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1305,9 +1382,11 @@ static void *_gnrc_netif_thread(void *args)
     while (1) {
-        DEBUG("gnrc_netif: waiting for incoming messages\n");
-        msg_receive(&msg);
+        msg_t msg;
```
I reduced the scope of the msg variable, it was declared at the top before.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1305,9 +1382,11 @@ static void *_gnrc_netif_thread(void *args)
-        DEBUG("gnrc_netif: waiting for incoming messages\n");
-        msg_receive(&msg);
+        msg_t msg;
+        _process_events_await_msg(netif, &msg);
```
Maybe add a comment on top, that `msg` will be set, when `_process_events_await_msg()` returns. It wasn't clear to me immediately.

gebart replied:
will do

gebart replied:
addressed

gebart commented:
@miri64 new version pushed. The block you mentioned was rewritten into a separate function, also the behaviour was slightly altered to let the thread check for events between each successive IPC message, even when more than one msg have been queued. This should make the behaviour more predictable and robust. Events are always processed as soon as possible, and any events which occur as a side effect of one of the processed messages will be handled before the next message is processed.

miri64 commented:
Ok, finally before I give my ACK, some doc on the new behavior would be nice (also mention that events take precedence over messages and ISR events are handled as a event)... I don't know exactly this should go though :-/. Maybe in a newly introduced details section in `net/gnrc/netif.h`?

miri64 commented on sys/include/net/gnrc/netif.h:
```diff
@@ -75,6 +78,22 @@ typedef struct {
+     * gnrc_netif thread, instead of having a single global instance which will
+     * not work if the system has more than one network interface.
+     * The _event_cb function of gnrc_netif.c uses this pointer to be able to
+     * post events from interrupt context.
```
It isn't really clear to me (without looking at the code), what you mean by "allocate ISR events"... How can one allocate an event? Yes, I know that you mean an instance of `event_t` but without any knowledge about that this might be difficult to grasp.

miri64 commented on sys/include/net/gnrc/netif.h:
```diff
@@ -75,6 +78,22 @@ typedef struct {
+     * gnrc_netif thread, instead of having a single global instance which will
+     * not work if the system has more than one network interface.
+     * The _event_cb function of gnrc_netif.c uses this pointer to be able to
+     * post events from interrupt context.
```
Different topic: why don't you just expose the type and allocate it here in full instead of a pointer?

miri64 replied:
Ok, the current comment makes it clearer.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -40,6 +40,16 @@
+typedef struct {
+    event_t super;
+    netdev_t *dev;
+} event_netdev_t;
```
If you don't want to make this type public, I'd prefer to prefix it with `_` to make the usage code easier to follow.

miri64 commented:
If this PR is rebased (and squashed) and https://github.com/RIOT-OS/RIOT/pull/9326#discussion_r235693209 is addressed optionally, I think we can finally merge this (sorry dropped off my radar again >.<)

miri64 commented:
@gebart ping?

miri64 commented:
Ping @gebart?

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1339,9 +1417,13 @@ static void *_gnrc_netif_thread(void *args)
         /* dispatch netdev, MAC and gnrc_netapi messages */
+        DEBUG("gnrc_netif: message %u\n", (unsigned)msg.type);
         switch (msg.type) {
             case NETDEV_MSG_TYPE_EVENT:
```
I believe this case can be `#ifndef`'d

jia200x commented:
@gebart this is indeed an interesting feature! Thanks

Since there's no activitry for some time, mind if I take it?

miri64 commented:
Adopted by @bergzand in #13669.
